lib/libmbedtls/core/sm2-dsa.c

```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019-2021 Huawei Technologies Co., Ltd
+ */
+
+#include <compiler.h>
+#include <crypto/crypto.h>
+#include <mbedtls/bignum.h>
+#include <mbedtls/ecp.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <util.h>
+#include <utee_defines.h>
+
+#include "sm2-dsa.h"
+
+/* SM2 uses 256 bit unsigned integers in big endian format */
+#define SM2_INT_SIZE_BYTES 32
+
+static int rng_read(void *ignored __unused, unsigned char *buf, size_t blen)
+{
+	if (crypto_rng_read(buf, blen))
+		return MBEDTLS_ERR_MPI_FILE_IO_ERROR;
+	return 0;
+}
+
+/* Generate random number 1 <= n < max */
+static TEE_Result gen_random_upto(mbedtls_mpi *n, mbedtls_mpi *max)
```

Comment:
This looks very much like the one in `lib/libmbedtls/core/sm2-pke.c`. Perhaps a shared implementation can be used instead?

Reply:
It is totally identical indeed. I will move it to a helper file.